File: src/Orleans.Security/GrainWithClaimsPrincipal.cs
using System.Security.Claims;
using System.Threading.Tasks;
using Orleans.Runtime;

namespace Orleans.Security
{
    public class GrainWithClaimsPrincipal : Grain, IIncomingGrainCallFilter
    {
        protected ClaimsPrincipal User;
        
        public async Task Invoke(IIncomingGrainCallContext context)
        {
            User = (ClaimsPrincipal)RequestContext.Get(ConfigurationKeys.ClaimsPrincipalKey);
            
            await context.Invoke();
        }
    }
}

File: src/Orleans.Security/GrainWithClaimsPrincipal.fs
namespace Orleans.Security

//open System.Threading.Tasks
//open Orleans.Runtime
//open System.Security.Claims
//open Orleans

//// TODO: Activation access violation. A non-activation thread attempted to access activation services
//type GrainWithClaimsPrincipal()=
//    [<DefaultValue>] val mutable User : ClaimsPrincipal 
//    inherit Orleans.Grain
//    interface IIncomingGrainCallFilter with
//        member this.Invoke(context: IIncomingGrainCallContext) = 
//            async {
//                this.User <- RequestContext.Get(ConfigurationKeys.ClaimsPrincipalKey) :?> ClaimsPrincipal
//                do! context.Invoke() |> Async.AwaitTask
//            } |> Async.StartAsTask :> Task